### PR TITLE
coqPackages.mathcomp: 1.17.0 -> 1.18.0

### PR DESCRIPTION
**Pull request by proux01**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/coq-modules/mathcomp/default.nix [hunk 1]
```diff
@@ -19,9 +19,10 @@ let
   owner = "math-comp";
   withDoc = single && (args.withDoc or false);
   defaultVersion = with versions; lib.switch coq.coq-version [
-      { case = isGe "8.15"; out = "1.17.0"; }
+      { case = isGe "8.16"; out = "1.18.0"; }
       { case = range "8.16" "8.18"; out = "2.1.0"; }
       { case = range "8.16" "8.18"; out = "2.0.0"; }
+      { case = range "8.15" "8.18"; out = "1.17.0"; }
```
Comment (vbgl) [inline pkgs/development/coq-modules/mathcomp/default.nix:25]: I think that this line should be moved two lines up.

Comment (proux01) [reply pkgs/development/coq-modules/mathcomp/default.nix:25]: Why not but isn't it equivalent?

Comment (vbgl) [reply pkgs/development/coq-modules/mathcomp/default.nix:25]: Hmm. You might be right…

File: pkgs/development/coq-modules/graph-theory/default.nix [hunk 2]
```diff
@@ -1,21 +1,25 @@
...
 
   inherit version;
-  defaultVersion = with lib.versions; lib.switch coq.coq-version [
-    { case = range "8.14" "8.16"; out = "0.9.1"; }
-    { case = range "8.12" "8.12"; out = "0.9"; }
+  defaultVersion = with lib.versions; lib.switch [ coq.coq-version mathcomp.version ] [
+    { cases = [ (isGe "8.16") (range "2.0.0" "2.1.0") ]; out = "0.9.3"; }
+    { cases = [ (range "8.14" "8.18") (range "1.13.0" "1.18.0") ]; out = "0.9.2"; }
+    { cases = [ (range "8.14" "8.16") (range "1.13.0" "1.14.0") ]; out = "0.9.1"; }
+    { cases = [ (range "8.12" "8.13") "1.12.0" ]; out = "0.9"; }
```
Comment (vbgl) [inline pkgs/development/coq-modules/graph-theory/default.nix:19]:
```suggestion
    { cases = [ (range "8.12" "8.13") (range "1.12.0" "1.14.0") ]; out = "0.9"; }
```

**Review by vbgl — COMMENTED** [review]
The default version of `mathcomp-infotheo` might be generalized to take into account the update of `mathcomp-analysis`, e.g.,

~~~
--- a/pkgs/development/coq-modules/mathcomp-infotheo/default.nix
+++ b/pkgs/development/coq-modules/mathcomp-infotheo/default.nix
@@ -7,7 +7,7 @@ mkCoqDerivation {
   inherit version;
   defaultVersion = with lib.versions; lib.switch [ coq.version mathcomp-analysis.version] [
     { cases = [ (isGe "8.17") (isGe "0.6.0") ];                  out = "0.5.2"; }
-    { cases = [ (range "8.15" "8.16") (range "0.5.4" "0.6.2") ]; out = "0.5.1"; }
+    { cases = [ (range "8.15" "8.16") (range "0.5.4" "0.6.5") ]; out = "0.5.1"; }
   ] null;
   release."0.5.1".sha256 = "sha256-yBBl5l+V+dggsg5KM59Yo9CULKog/xxE8vrW+ZRnX7Y=";
   release."0.5.2".sha256 = "sha256-8WAnAV53c0pMTdwj8XcUDUkLZbpUgIQbEOgOb63uHQA=";
~~~

File: pkgs/development/coq-modules/graph-theory/default.nix (hunk 2 above)
Comment (proux01) [reply pkgs/development/coq-modules/graph-theory/default.nix:19]: Are you sure this works? OPAM only lists 1.12 and the release messages on https://github.com/coq-community/graph-theory/releases reads:
* 0.9: "This release is compatible with mathcomp-1.12"
* 0.9.1: "This is a maintenance release for mathcomp-1.13 and mathcomp 1.14."

**Comment by proux01** [comment]
> The default version of `mathcomp-infotheo` might be generalized to take into account the update of `mathcomp-analysis`,

Indeed, analysis is currently waiting for a release working on MC 1.18 (master works it's mostly a matter of releasing), so let's see what happens then.

File: pkgs/development/coq-modules/graph-theory/default.nix (hunk 2 above)
Comment (vbgl) [reply pkgs/development/coq-modules/graph-theory/default.nix:19]: This is what was here before your PR.

Comment (proux01) [reply pkgs/development/coq-modules/graph-theory/default.nix]: Indeed

**Comment by proux01** [comment]
@vbgl added Analysis 0.6.6 compatible with MC 1.18, successfully tested there: https://github.com/coq-community/coq-nix-toolbox/pull/177

**Comment by vbgl** [comment]
My comment about infotheo still applies.

**Comment by proux01** [comment]
@vbgl indeed, thanks, done (and successfully tested there https://github.com/coq-community/coq-nix-toolbox/pull/177 )
Unfortunately, this means we have to give up MC 1.18 (and remains on 1.17) on Coq 8.16 but I guess it's a reasonnable trade-offf.

**Comment by vbgl** [comment]
Many thanks!

**Comment by proux01** [comment]
Thanks

**Pull request merged**